draft/XXXX-table-metadata/README.md

```diff
@@ -0,0 +1,80 @@
+
+## Motivation
+
+The main motivation for this is storing type information for each column in a table. We now have at least two use cases for storing this information.
+
+1. Basic computation/comparison. If we attempt to use these tables for creating graphs and deriving useful data, it's necessary that we store type information, so that we are able to properly compare and compute certain values. For example, if you have a column of numbers, but we store them as strings (the current behavior), then if you try to do comparisons of these column values between different entries, you either will be doing incorrect string comparisons, or will simply not be able to do the comparisons/computations you need to.
+
+2. Providing this information to other applications, so that they do not have to do this data inspection themselves, or if this information is required by those applications. Specifically, [UpSet](https://github.com/VCG/upset) is the first application for which this will be used.
+
+## Proposals
+
+#### Data Model
+There are several way we could go about this, but in my opinion, the cleanest way is to store this information in a special collection, in the workspace that the table resides in. This special collection would be named something like `multinet_meta`, and each entry would correspond to the metadata on a specific graph or table. I think the data model would look something like this:
+
+
+```
+{
+  item_id: str                 // The ID of the table/graph this data belongs to
+  item_type: "table" | "graph" // The major type of the item
+  table?: {                    // Only present if the major type is "table"
+    type: "edge" | "node"      // The table type
+    columns: [
+      {
+        key: str               // E.g. {"foo": "bar"}, "foo" is the key
+        type: "number" | "label" | "category" | "date" | "bool"
+      }
+    ]
+  }
+  graph?: {                    // Only present if the major type is "graph"
+    // TBD
+  }
+
+}
+```
+
+#### API
+There are two main operations that are necessary for this functionality.
+
+1. Specifying the metadata on table creation
+2. Retrieving this metadata
+
+Ideally, our API would allow for the creation of a table/graph record, with more versatile options for specifying metadata. However, our API isn't currently structured this way. Instead, in order to conform to our existing API, we would likely just add this metadata object as a query parameter onto each upload endpoint. For retrieving this metadata, the API should resemble the following:
+
+```
+GET /api/workspaces/{workspace}/tables/{table}/metadata
+```
+
+##### Specifying Metadata on Table Creation
+
+The data model of the API would differ slightly from that of what we store internally, to allow for a single request to describe metadata on multiple entities (mainly in the case of graphs. The data format of this API may resemble the following:
```

Comment:
I'm not sure why we would need to update metadata on multiple things at once. Generally, tables are uploaded, then networks are created by selecting tables. At the time a network is created, I can see having some metadata also set up to be set on it. If you're thinking of the case where a data format automatically creates tables and a network, even in that case it seems like multiple calls to set metadata on different objects would be simpler to implement. What do you think?

Reply:
With our current API, we would need to pass the type information of both node and edge tables along at graph creation time, since the upload/creation is a synchronous call, and it needs this information _before_ creating the tables/graphs, as this will affect the data type that is chosen to store each column.

Reply:
But in general, the tables are created first, then a network is created. The type information would already be there in the table metadata. Right?

But I think it could be different for the special format network creation routines, like the ones for nested JSON and D3-formatted graph data. Is that what you're referring to here?

Reply:
Yes, this is specifically for uploading a graph directly (D3 JSON, Nested JSON or Newick). In these cases, the tables (and subsequent graph) will be created synchronously in the api call, so we would need to provide this metadata as a part of the API call which creates them.

Reply:
We'll need to figure out UI/UX for pre-analyzing the data tables and allowing the user to specify types. But your approach makes sense to me now.